### ISSUE TITLE
Core 2.0 multiple auth schemes problem: default one overrides all, and no default ignores all

### ISSUE THREAD

**Issue by vgribok** [aspnet/Security#1469]
The problem I am reporting seems to be similar to https://github.com/aspnet/Security/issues/1381.

# Summary
My TL;DR summary is that in ASP.NET Core 2,0 it seems impossible to have multiple authentication schemes handling requests with corresponding values of the `Authorization` header, while we had no trouble implementing this in 1.1 and earlier with OWIN.

# Details
Using ASP.NET Core 2.0, we are adding two custom authentication schemes, let's call them "Abc" and "Xyz", using `services.AddSchema()` method. When I send a request with `Authorization` header having either `Abc SomeAuthInfo` or `Xyz SomeOtherAuthInfo` value, I expect a corresponding handler to be called by the framework, as it happened in OWIN and later in Core 1.1. That does not happen. The only non-invasive way to make it work was specifying default schema via `services.AddAuthentication("Abc")`, which resulted in "Abc" scheme working and "Xyz" not working. 

Eventually, I have to do this:
```
di.AddScoped<IAuthenticationHandlerProvider, AuthSchemeMappingAuthenticationHandlerProvider>();
di.AddSingleton<IAuthenticationSchemeProvider, BugWorkaroundAuthenticationSchemeProvider>();
```
with those two hand-rolled classes hacking the default ASP.NET Core 2.0 logic letting us look at what Authorization header value starts with - the scheme - and map it the appropriate handler.

# The Question
My question is, is this a bug, or have I simply failed to find a legitimate way to map auth schemes to their handlers?

**Comment by sschleicher208** [comment]
I have this exact same issue.  I have Jwt and Facebook authentication enabled (Jwt is default scheme).  When I pass Jwt token in the authorization header, works perfectly.  When I pass the access token in the authorization header that I received from facebook, it fails.  Did you ever find anything out about this?

**Comment by vgribok** [comment]
So far I have not heard anything from ASP.NET dev team. I have implemented a workaround by replacing two classes as shown in my post, but I hope to replace it with a better solution. I'm hoping someone can suggest a better way.

**Comment by PinpointTownes** [comment]
> My question is, is this a bug, or have I simply failed to find a legitimate way to map auth schemes to their handlers?

Good or bad, this is a deliberate change that was announced here: https://github.com/aspnet/Announcements/issues/262.
Basically, the new stack only allows you to have a single automatic scheme handler by default. Having to override the scheme provider is expected if you want to change the default selection mechanism.

Don't miss this questions/answers thread if you need more information: https://github.com/aspnet/Security/issues/1338.

**Comment by davidfowl** [comment]
@vgribok What @PinpointTownes says is right, this is by design. However, nothing stops you from calling context.AuthenticateAsync("Scheme") to access the right user information.

How were you doing this before? Was it dynamic basic on something in the incoming request?

**Comment by vgribok** [comment]
Yes, it was dynamic based on the scheme name in the incoming requests' `Authorization` header. In Core 1.1 it worked without us having to manually do the mapping - we just specified every scheme as default, which was weird but possible.

What would be helpful is a clean code example showing how to do it properly in Core 2.0. Right now it's not hard to figure out how to add multiple schemes, but getting them to be employed is hard, and no I could not find clear guidance. "Programmer's Guide" type of docs would be very appreciated.

**Comment by davidfowl** [comment]
The code in 2.0 would look like a piece of middleware that does this:


```c#
app.UseAuthentication();

app.Use(async (context, next) =>
{
    // Write some code that determines the scheme based on the incoming request
    var scheme = GetSchemeForRequest(context);
    var result = await context.AuthenticateAsync(scheme);
    if (result.Succeeded)
    {
        context.User = result.Principal;

    }
    await next();
});
```

FWIW, we're actively looking at some first class improvements in this area for 2.1 (see https://github.com/aspnet/Security/pull/1410).

**Comment by vgribok** [comment]
Thanks.
Yes, we are using MVC, but we build same kind of identity & principal from three and potentially more different auth schemes. Our auth handlers have federation-like functionality, and because of that our roles will be the same regardless of the scheme type in the inbound request, which in turn makes explicit usage of scheme with controllers impossible.

**Comment by vgribok** [comment]
I'd like to add 2c on what would be intuitive design for me. I don't quite understand the value of the DefaultAuthenticationScheme. (Default challenge scheme I understand.) IMHO, the process here is two-step: 1) Building Principal from request, and 2) authorizing protected resources (like Controller methods) by matching Claims supplied by the Principal to ones demanded by the protected resource via `Authorize` attribute. In this flow, at step 1, multiple sources of identity and claims can be supplied - via potentially multiple `Authorization` headers, cookies, and wherever else is in the request. The only value of a schema I can see is to simplify mapping of "properly"-supplied auth info, meaning `Authorization` header to handlers. That, I hoped, was implemented in Core 2.0 and allowed to get rid of per-schema middleware we had to have in Core 1.1 and OWIN. So far very good: having only one framework-supplied `AuthenticationMiddleware` may have even more benefits, like maybe calling multiple auth handlers in parallel, for example, bypassing the sequential nature of a middleware we had to do in 1.1. But in this design, for the life of me, I can't see the place for the default auth schema. 

I _may_ be able to see default auth schema only in the case where Principal has multiple claim types, and `Authorize` attribute does not specify `Claim.Type` (`AuthorizeAttribute.AuthenticationSchemes`) making it impossible to determine what `Roles` property is really specifying. There default auth schema, which would really be default claim type, would make sense to me. Current design, though, with AuthorizeAttribute.AuthenticationSchemes implying relationship to the Claim.Type, while not being it IFAICT, is a stretch: as I mentioned, a single incoming Authorization header with a single scheme can result in a single IIdentity being a part of Principal with multiple types of claims. Just as true is the observation that multiple Authorization headers in the same request can result in a Principal having Claims of just a single Claim.Type. If you agree with that assumption, it effectively breaks default scheme as a coupling link between the inbound request and the application of claims authorization.

To summarize, current approach of somewhat tight coupling via default scheme of auth info in the inbound requests with applying Claims-based authorization is a stretch. IMHO, there should not be coupling at all, and maybe that's what can be done using Policies - I still need to learn that part, but that adds new layers of complexity and to the length of a learning curve. Even back in 2013 when .NET 5.2 introduced Claims authorization, it was too complex to most folks used to ASP.NET cookie auth, but what I am seeing now it really complex, while not seemingly covering relatively straightforward use cases.

**Comment by HaoK** [comment]
So the DefaultAuthenticateScheme's purpose is twofold, its mainly used by the single default authentication middleware to set HttpContext.User to the result of a specific authentication scheme.  

Its fine to have multiple sources of identity and claims used to produce the HttpContext.User that you then use for Authorization.  You can either write a custom uber AuthenticationHandler and register that as your default scheme, and it can pull from whatever sources you want.  Or you can add your own custom middleware that assembles the httpContext.User however you like, and just don't use DefaultAuthenticationScheme.

Authentication today isn't aware of authorization at all, it only cares about producing a `ClaimsPrincipal` per scheme.  Authorization has to be aware of authentication somewhat, so AuthenticationSchemes in the policy is a mechanism for you to associate the policy with schemes used to build the effective claims principal for authorization (or it just uses the default httpContext.User for the request, which does rely on DefaultAuthenticateScheme)

**Comment by davidfowl** [comment]
> Or you can add your own custom middleware that assembles the httpContext.User however you like, and just don't use DefaultAuthenticationScheme.

This is what my sample does above, but what breaks down is the call to Challenge/Forbid if there is no user. The new authentication schemes solves this nicely.

> Its fine to have multiple sources of identity and claims used to produce the HttpContext.User that you then use for Authorization. You can either write a custom uber AuthenticationHandler and register that as your default scheme, and it can pull from whatever sources you want. 

This is a super interesting idea and not one I thought about before. I might try to whip up a sample with this.

**Comment by HaoK** [comment]
Yeah this is basically using the authenticationScheme as the authentication policy today

**Comment by brockallen** [comment]
>> Its fine to have multiple sources of identity and claims used to produce the HttpContext.User that you then use for Authorization. You can either write a custom uber AuthenticationHandler and register that as your default scheme, and it can pull from whatever sources you want.

>> This is a super interesting idea and not one I thought about before. I might try to whip up a sample with this.


This is what we do in our IdentityServer access token validation handler that is implemented in terms of other handlers (specifically JWT Bearer and introspection).

https://github.com/IdentityServer/IdentityServer4.AccessTokenValidation

**Comment by vgribok** [comment]
> Its fine to have multiple sources of identity and claims used to produce the HttpContext.User that you then use for Authorization. You can either write a custom uber AuthenticationHandler and register that as your default scheme, and it can pull from whatever sources you want. Or you can add your own custom middleware that assembles the httpContext.User however you like, and just don't use DefaultAuthenticationScheme.

For sure, I can certainly roll out my own anything, but with `services.AddScheme()` enabling multiple schemes and with the presence of a general-purpose `app.UseAuthentication()`, the requirement to add custom components to make use of multiple schemes is surprising, as the framework seems to have all necessary information to make scheme-handler mapping generic and transparent, improving user experience by not requiring users create custom components. That's what my custom AuthenticationHandlerProvider and AuthenticationSchemeProvider do together, without resorting to using custom middleware or the uber-handler (unless handler provider is considered to be that): they map scheme from the inbound Authorization header to a registered handler with matching scheme, without needing to know what those schemes and handlers are. I am basically asking for the less hack-y way to do it (complete example would be more helpful than an outline), and to understand what is the reason why it is not a part of the framework, i.e. whether it's just a matter of ASP.NET team having time to do it, or if there is a reason for current approach to stay like this forever.

**Comment by davidfowl** [comment]
> Yeah this is basically using the authenticationScheme as the authentication policy today

@HaoK how hacky is something like this? 

```C#

services.AddAuthentication("Dynamic")
        .AddJwtBearer("Scheme2", options =>
        {
        })
        .AddJwtBearer("Scheme1", options =>
        {
        })
        .AddDynamicAuthentication(options =>
        {
            options.SchemeSelector = context =>
            {
                // Write whatever logic you need to pick the "current" scheme
                string authHeader = context.Request.Headers["Authorization"];
                // Authorization: Scheme 
                var scheme = authHeader.Split(' ')[0];

                return scheme;
            };
        });
```

I guess the authN policy concept is still cleaner because some authN policies require a unique set of schemes for (authenticate, challenge, signin, signout). For example, a "facebook" policy would look like:

"Facebook" -> (authenticate: "cookies", challenge: "facebook", signin: "cookies", signout: "cookies")

**Comment by vgribok** [comment]
Thanks! That's not hacky, and I'll happily use it. I hope this will become the default behavior in one of the upcoming versions of ASP.NET Core.

**Comment by HaoK** [comment]
@davidfowl that looks fine and no hacky really.

So thinking about it more, you really could implement a scheme as a policy today already.

Just have the handler forward to the right other schemes already in its implementation.  For your example: lets say we added a AuthenticationPolicyHandler and just made it easy to setup schemes that behaved as policies (forwarders):

You'd still do something 
`.AddAuthenticationPolicy("FacebookPolicy", policy => {
      policy.DefaultScheme = "Cookies"
      policy.ChallengeScheme = "Facebook";
 })`

And we'd have a new base `AuthenticationPolicyHandler` that could just take a named AuthenticationPolicyOption "FacebookPolicy" and used that to dispatch the Authenticate/Challenge/Forbid/SignIn/SignOut to the appropriate schemes.

Might be cleaner/simpler than building out all the additional services/classes in the current PR for authN policies.

And probably instead of calling it authentication policy, call it something else to disambiguate it from AuthorizationPolicy. (AuthenticationForwarderHandler or something)

**Comment by davidfowl** [comment]
> Might be cleaner/simpler than building out all the additional services/classes in the current PR for authN policies.

Yep that's what I'm thinking as well.

I was thinking that you could get all the flexibility making the scheme selector a policy selector.

**Comment by Eilon** [comment]
Closing this issue because we're using https://github.com/aspnet/Security/issues/1359 as the main discussion (and there's a linked PR as well). If there's anything not covered there, please comment in the other thread. Thanks!

**Comment by EvAlex** [comment]
Thank you, @davidfowl ! Your comment helped me to solve another issue, that I stumbled upon while implementing IdentityServer4 with REST API controllers in the same app. Existing samples show how to implement this as separate apps: one with IdentityServer4 for auth and other apps with authentication based on JWT in request headers via AuthorizeAttribute. The problem arose because using `services.AddAuthentication(...).AddJwtBearer(...)` together with `services.AddIdentityServer(...)` leads to incorrect behavior: logging in didn't actually log in. But without `services.AddAuthentication(...).AddJwtBearer(...)` my AuthorizeAttribute didn't handle JWT token in Authorization header. So I implemented it manually:

```cs
app.Use(async (context, next) => 
    await AuthenticateApiRequestByAuthorizeHeader(context, next, app.ApplicationServices));
```

```cs
private async Task AuthenticateApiRequestByAuthorizeHeader(HttpContext context, Func<Task> next, IServiceProvider serviceProvider)
{
    bool isApiRequest = context.Request.Path.StartsWithSegments("/api");
    if (isApiRequest)
    {
        string token = context.Request.Headers["Authorization"].FirstOrDefault()?.Substring("Bearer ".Length);
        if (token != null)
        {
            var validator = serviceProvider.GetServiceOrDie<ITokenValidator>();
            var validationResult = await validator.ValidateAccessTokenAsync(token);

            if (!validationResult.IsError && validationResult.Claims.Any(c => c.Type == "email"))
            {
                string email = validationResult.Claims.FirstOrDefault(c => c.Type == "email")?.Value;

                var userManager = serviceProvider.GetServiceOrDie<UserManager<ApplicationUser>>();
                var user = await userManager.FindByEmailAsync(email);

                if (user != null)
                {
                    var signInManager = serviceProvider.GetServiceOrDie<SignInManager<ApplicationUser>>();
                    await signInManager.SignInAsync(user, isPersistent: false);

                    var userClaimsPrincipalFactory = serviceProvider.GetServiceOrDie<UserClaimsPrincipalFactory<ApplicationUser>>();
                    context.User = await userClaimsPrincipalFactory.CreateAsync(user);
                }
            }
        }
    }
    await next();
}
```

where GetServiceOrDie is an extension
```cs
public static T GetServiceOrDie<T>(this IServiceProvider serviceProvider)
{
    var service = serviceProvider.GetService<T>();
    if (EqualityComparer<T>.Default.Equals(service, default(T)))
        throw new Exception($"Failed to get service of type {typeof(T)}");
    return service;
}
```

**Comment by AKlaus** [comment]
**Problem**
I had a similar problem, when I was implementing a middleware with an HTTP protocol extension. And to work via that protocol, a custom cookie authentication is required. It all wouldn't be a problem if it was a separate app server, but the requirement was to run under a specific path on a server, which is using a different authentication scheme. 
So the problem is that I can't easy specify a custom rule like `context.Request.Path.Value.StartsWith("/documents")` in the `Startup.cs` for applying an authentication scheme. And I needed it because the `/documents` path was handled in a middleware, not in a controller.

**Workaround**
A workaround, which worked for me was to write my own implementation of [IAuthenticationHandlerProvider](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authentication.iauthenticationhandlerprovider) (with just a minor difference from the [original implementation](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authentication.authenticationhandlerprovider)), which would resolve the authentication scheme differently for different parts in [GetHandlerAsync()](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authentication.authenticationhandlerprovider.gethandlerasync) method.
The remaining bit was to register my implementation in [IStartup.ConfigureServices()](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.hosting.istartup.configureservices?view=aspnetcore-2.0#Microsoft_AspNetCore_Hosting_IStartup_ConfigureServices_Microsoft_Extensions_DependencyInjection_IServiceCollection_) method like the OP showed:
```
di.AddScoped<IAuthenticationHandlerProvider, AuthSchemeMappingAuthenticationHandlerProvider>();
```
It feels hacky, but works.

**Questions**
- @vgribok, why did you need your own [IAuthenticationSchemeProvider](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authentication.authenticationschemeprovider) implementation? It seems working without it.
- Is it even possible to implement a dynamic scheme, which can pass authentication on other schemes (as suggested by @davidfowl above)? Apart from .NET Core v1 implementation via middleware, new `AuthenticationHandler` classes can't pass it on to the next one.

**Comment by saborrie** [comment]
@AKlaus 
>Is it even possible to implement a dynamic scheme, which can pass authentication on other schemes (as suggested by @davidfowl above)? Apart from .NET Core v1 implementation via middleware, new AuthenticationHandler classes can't pass it on to the next one.

I don't know much about this but while reading the code in this link, I think I might have found out how you pass on auth to other schemes: https://joonasw.net/view/creating-auth-scheme-in-aspnet-core-2 

```
protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
{
    ...
    return AuthenticateResult.NoResult();
}
```

**Comment by speak2san** [comment]
I have a API which is configured to be authenticated from "Azure AD B2C" token. Now I also need to make it accept "Azure AD" token. Can someone please help with the clear direction ?

**Comment by webbkvalite** [comment]
Here's the code I use to add JWT Authentication (or any other number of authentications) automatically to all controllers via an attribute to a base controller:

```
    public class LoadJwtAttribute : Attribute, IFilterFactory
    {
        public bool IsReusable => true;

        public IFilterMetadata CreateInstance(IServiceProvider serviceProvider)
        {
            return new LoadJwtFilter();
        }
    }

    public class LoadJwtFilter : IAsyncAuthorizationFilter
    {
        public async Task OnAuthorizationAsync(AuthorizationFilterContext context)
        {
            var authenticateJwt = await context.HttpContext.AuthenticateAsync(JwtBearerDefaults.AuthenticationScheme);
            if (authenticateJwt.Succeeded)
            {
                context.HttpContext.User = authenticateJwt.Principal;
            }
        }
    }
```

**Comment by brewsoftware** [comment]
That last post misses the point. How do you add 2 challenge schemes.

i.e. JwtBearer + SecretHeader. How they map to a user/permission is might also need to be customised.

If a request comes from a user and has a bearer token then use the JWT Bearer Auth, if it comes from a monitoring service use the SecretHeader and assign the appropriate permissions.

Currently it doesn't seem to have a way of handling this? At the moment I've rolled us back to using attributes of [AllowAnonymouse] and [CustomAuthAttribute] to get around the limitation. Really annoying thing is that we can't use the same call for the two different authentication systems..

**Comment by openidauthority** [comment]
@brewsoftware, I think in 2.1 you could do something like this (not tested):
```
            services.AddAuthentication(options =>
            {
                options.DefaultAuthenticateScheme = "smart";
                options.DefaultChallengeScheme = "smart";
            })
            .AddPolicyScheme("smart", "Bearer or Secret Header", options =>
            {
                options.ForwardDefaultSelector = context =>
                {
                    var authHeader = context.Request.Headers["Authorization"];
                    if (authHeader?.StartsWith("Bearer ") == true)
                    {
                        return JwtBearerDefaults.AuthenticationScheme;
                    }
                    return "SecretHeader";
                };
            })
            .AddJwtBearer(JwtBearerDefaults.AuthenticationScheme, options =>
            {
                options.Authority = Configuration["Authentication:Authority"];
                options.Audience = Configuration["Authentication:Audience"];
            })
            ... // add scheme for SecretHeader etc.
```

**Comment by Maximys** [comment]
Can anybody explain me next thing:
I have next ConfigureServices method:
```
public void ConfigureServices(IServiceCollection services)
        {
            services.AddDbContext<ApplicationContext>(options =>
                options.UseSqlServer(Configuration.GetConnectionString("DefaultConnection")));

            services.AddIdentity<User, IdentityRole>()
                    .AddEntityFrameworkStores<ApplicationContext>()
                    .AddErrorDescriber<MyIdentityErrorDescriber>()
                    .AddDefaultTokenProviders()
                    .AddPasswordValidator<UsernameAsPasswordValidator<User>>();
            services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
                .AddJwtBearer(options =>
                {
                    options.TokenValidationParameters = new TokenValidationParameters()
                    {
                        ValidIssuer = Configuration[AccountController.TokenIssuerConfigName],
                        ValidAudience = Configuration[AccountController.TokenAudienceConfigName],
                        IssuerSigningKey = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(Configuration[AccountController.SymmetricSecurityKeyConfigName]))
                    };
                });


            services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_1);
        }
```

and next Configure method:
```
public void Configure(IApplicationBuilder app, IHostingEnvironment env)
        {
            app.UseDeveloperExceptionPage();

            app.UseStaticFiles();

            app.UseAuthentication();

            app.Use(async (context, next) =>
            {
                // Write some code that determines the scheme based on the incoming request
                var scheme = GetSchemeForRequest(context);
                var result = await context.AuthenticateAsync(scheme);
                if (result.Succeeded)
                {
                    context.User = result.Principal;

                }
                await next();
            });

            app.UseMvc(routes =>
            {
                routes.MapRoute(
                    name: "default",
                    template: "{controller=Home}/{action=Index}/{id?}");
            });
        }

        private string GetSchemeForRequest(HttpContext context)
        {
            //JUST FOR TEST
            return "Identity.Application";
        }
```
When I trying to execute next method of controller as unauthorized user from browser:
```
[Authorize(AuthenticationSchemes = JwtBearerDefaults.AuthenticationScheme)]
        [Authorize]
        public IActionResult Contact()
        {
            ViewData["Message"] = "Your contact page.";

            return View();
        }
```
I have only:

> "Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 GET http://localhost:44360/Home/Contact  
> Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Route matched with {action = "Contact", controller = "Home"}. Executing action WebApplication3.Controllers.HomeController.Contact (WebApplication3)
> 'dotnet.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App\2.1.0\Microsoft.IdentityModel.Tokens.dll'. Cannot find or open the PDB file.
> 'dotnet.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App\2.1.0\System.IdentityModel.Tokens.Jwt.dll'. Cannot find or open the PDB file.
> 'dotnet.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\2.1.0\System.Xml.ReaderWriter.dll'. Symbols loaded.
> 'dotnet.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App\2.1.0\Microsoft.IdentityModel.Logging.dll'. Cannot find or open the PDB file.
> 'dotnet.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App\2.1.0\Microsoft.IdentityModel.Protocols.OpenIdConnect.dll'. Cannot find or open the PDB file.
> Microsoft.AspNetCore.Authorization.DefaultAuthorizationService:Information: Authorization failed.
> Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Authorization failed for the request at filter 'Microsoft.AspNetCore.Mvc.Authorization.AuthorizeFilter'.
> Microsoft.AspNetCore.Mvc.ChallengeResult:Information: Executing ChallengeResult with authentication schemes (Bearer).
> Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler:Information: AuthenticationScheme: Bearer was challenged.
> Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executed action WebApplication3.Controllers.HomeController.Contact (WebApplication3) in 110.9756ms"

and 401 as result, but when I remove:

```
services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
                .AddJwtBearer(options =>
                {
                    options.TokenValidationParameters = new TokenValidationParameters()
                    {
                        ValidIssuer = Configuration[AccountController.TokenIssuerConfigName],
                        ValidAudience = Configuration[AccountController.TokenAudienceConfigName],
                        IssuerSigningKey = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(Configuration[AccountController.SymmetricSecurityKeyConfigName]))
                    };
                });
```

from ConfigureServices and

```
app.Use(async (context, next) =>
            {
                // Write some code that determines the scheme based on the incoming request
                var scheme = GetSchemeForRequest(context);
                var result = await context.AuthenticateAsync(scheme);
                if (result.Succeeded)
                {
                    context.User = result.Principal;

                }
                await next();
            });
```

from Configure, I see Login page as I expected.
As I can see, my problem and this bug are linked. How can I fix it?

**Comment by eopeter** [comment]
@openidauthority I tried this in 2.1 and works for api (jwt) and able to view website (cookie, facebook) with [AllowAnonymous] but I cannot login to website with cookie / facebook. When i enter credentials, I get routed back to login page. Any help is appreciated

**Comment by openidauthority** [comment]
@eopeter, maybe
[this code sample](https://github.com/leastprivilege/AspNetCoreSecuritySamples/blob/aspnetcore21/OidcAndApi/src/AspNetCoreSecurity/Startup.cs) will help?

**Comment by hiraldesai** [comment]
Question: How do I go about using Azure Active Directory authentication & client certificate authentication side by side? Since client authentication requires enabling a flag at IIS level (IIS express or Azure app service), wouldn't that means someone trying to pass bearer auth token in Authorization header will also need to provide a certificate?

**Comment by Tratcher** [comment]
You're correct, cert auth on IIS doesn't stack well with other types. This would work ok with HttpSysServer since it allows you to request a client certificate lazily.

**Comment by hiraldesai** [comment]
Thanks @Tratcher for confirming. I had initially thought of doing this with 2 auth providers and adding the auth type with scheme as a controller attribute but it looks like I'll have to split my project into multiple APIs.

I wonder if this is something that can be considered as an enhancement for future. Surely this isn't the right place for it, [is this the right place to log it?](https://windowsserver.uservoice.com/forums/310252-iis-and-web-server-role)

**Comment by saborrie** [comment]
@hiraldesai
> Thanks @Tratcher for confirming. I had initially thought of doing this with 2 auth providers and adding the auth type with scheme as a controller attribute but it looks like **I'll have to split my project into multiple APIs.**

Hopefully this is helpful for thatl: 
[http://dhickey.ie/2018/06/09/aspnet-core-nested-apps/](http://dhickey.ie/2018/06/09/aspnet-core-nested-apps/)

**Comment by stevenxi** [comment]
I have similar but not 100% same issues.
I have a project involves cookie auth and jwt token auth.
The idea is to use cookie for web page and web api, jwt for api only.

I've tried to use 2 auth schema same time in the auth policy builder. It works fine as authorization, but failed on challenge, probably because jwt auth overrides it. 

So at the end, I used 2 auth policies for the project. cookie only for web page. and cookie + jwt for the apis. It works fine now.

Code as below:
custom AuthroizeFilter, so I can set auth policy by route:

        private class CustomApiAuthroizeFilter : AuthorizeFilter
        {
            private readonly bool _possitive;
            public CustomApiAuthroizeFilter(AuthorizationPolicy policy, bool possitive)
                : base(policy)
            {
                _possitive = possitive;
            }

            public override Task OnAuthorizationAsync(AuthorizationFilterContext context)
            {
                var match = context.HttpContext.Request.Path.StartsWithSegments("/api", StringComparison.InvariantCulture);
                if (_possitive)
                {
                    if (match)
                        return base.OnAuthorizationAsync(context);
                }
                else
                {
                    if (!match)
                        return base.OnAuthorizationAsync(context);
                }
                return Task.CompletedTask;
            }
        }



API policy:

                    var apiPolicy = new AuthorizationPolicyBuilder()
                        .AddAuthenticationSchemes(JwtBearerDefaults.AuthenticationScheme, CookieAuthenticationDefaults.AuthenticationScheme)
                        .RequireAuthenticatedUser()
                        .Build();

                    o.Filters.Add(new CustomApiAuthroizeFilter(apiPolicy, true));

Page policy:

                    var pagePolicy = new AuthorizationPolicyBuilder()
                        .RequireAuthenticatedUser()
                        .Build();

                    o.Filters.Add(new CustomApiAuthroizeFilter(pagePolicy, false));

**Comment by tkueny-vtrax** [comment]
@openidauthority - that worked like a charm for me! The ForwardDefaultSelector on the CookieAuthenticationOptions works as it should. It allowed me to configure both cookie & JWT authorization based off the requesting url/route. I spent days on this, this is very much appreciated!
